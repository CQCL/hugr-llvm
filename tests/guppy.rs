use std::{env, path::PathBuf};

use rstest::{fixture, rstest};

struct TestConfig {
    python_bin: PathBuf,
    hugr_llvm_bin: PathBuf,
}

impl TestConfig {
    pub fn new() -> TestConfig {
        let python_bin = env::var("HUGR_LLVM_PYTHON_BIN")
            .map(Into::into)
            .ok()
            .or_else(|| pathsearch::find_executable_in_path("python"))
            .unwrap_or_else(|| panic!("Could not find python in PATH or HUGR_LLVM_PYTHON_BIN"));
        let hugr_llvm_bin = env!("CARGO_BIN_EXE_hugr-llvm").into();
        TestConfig {
            python_bin,
            hugr_llvm_bin,
        }
    }
}

#[fixture]
fn test_config() -> TestConfig {
    TestConfig::new()
}

#[rstest]
fn test_even_odd(test_config: TestConfig) {}
