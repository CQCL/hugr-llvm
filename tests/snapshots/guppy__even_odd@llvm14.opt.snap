---
source: tests/guppy.rs
info:
  program: hugr-llvm
  args:
    - "-"
---
success: true
exit_code: 0
----- stdout -----
; ModuleID = 'module'
source_filename = "module"

define { i32, {}, {} } @_hl.is_even.29(i64 %0) {
alloca_block:
  %1 = icmp eq i64 %0, 0
  %2 = select i1 %1, { i32, {}, {} } { i32 1, {} poison, {} undef }, { i32, {}, {} } { i32 0, {} undef, {} poison }
  %3 = extractvalue { i32, {}, {} } %2, 0
  %4 = extractvalue { i32, {}, {} } %2, 1
  %5 = extractvalue { i32, {}, {} } %2, 2
  %cond = icmp eq i32 %3, 1
  %6 = insertvalue { i64 } undef, i64 %0, 0
  %7 = insertvalue { i32, { i64 }, {} } { i32 0, { i64 } poison, {} poison }, { i64 } %6, 1
  %"07.0" = select i1 %cond, { i32, { i64 }, {} } { i32 1, { i64 } poison, {} undef }, { i32, { i64 }, {} } %7
  %8 = extractvalue { i32, { i64 }, {} } %"07.0", 0
  %cond29 = icmp eq i32 %8, 1
  br i1 %cond29, label %15, label %9

9:                                                ; preds = %alloca_block
  %10 = extractvalue { i32, { i64 }, {} } %"07.0", 1
  %11 = extractvalue { i64 } %10, 0
  %12 = sub i64 %11, 1
  %13 = call { i32, {}, {} } @_hl.is_odd.30(i64 %12)
  %14 = extractvalue { {} } undef, 0
  br label %18

15:                                               ; preds = %alloca_block
  %16 = extractvalue { i32, { i64 }, {} } %"07.0", 2
  %17 = extractvalue { {} } undef, 0
  br label %18

18:                                               ; preds = %9, %15
  %"02.0" = phi { i32, {}, {} } [ %13, %9 ], [ { i32 1, {} poison, {} undef }, %15 ]
  ret { i32, {}, {} } %"02.0"
}

define { i32, {}, {} } @_hl.is_odd.30(i64 %0) {
alloca_block:
  %1 = icmp eq i64 %0, 0
  %2 = select i1 %1, { i32, {}, {} } { i32 1, {} poison, {} undef }, { i32, {}, {} } { i32 0, {} undef, {} poison }
  %3 = extractvalue { i32, {}, {} } %2, 0
  %4 = extractvalue { i32, {}, {} } %2, 1
  %5 = extractvalue { i32, {}, {} } %2, 2
  %cond = icmp eq i32 %3, 1
  %6 = insertvalue { i64 } undef, i64 %0, 0
  %7 = insertvalue { i32, { i64 }, {} } { i32 0, { i64 } poison, {} poison }, { i64 } %6, 1
  %"07.0" = select i1 %cond, { i32, { i64 }, {} } { i32 1, { i64 } poison, {} undef }, { i32, { i64 }, {} } %7
  %8 = extractvalue { i32, { i64 }, {} } %"07.0", 0
  %cond29 = icmp eq i32 %8, 1
  br i1 %cond29, label %15, label %9

9:                                                ; preds = %alloca_block
  %10 = extractvalue { i32, { i64 }, {} } %"07.0", 1
  %11 = extractvalue { i64 } %10, 0
  %12 = sub i64 %11, 1
  %13 = call { i32, {}, {} } @_hl.is_even.29(i64 %12)
  %14 = extractvalue { {} } undef, 0
  br label %18

15:                                               ; preds = %alloca_block
  %16 = extractvalue { i32, { i64 }, {} } %"07.0", 2
  %17 = extractvalue { {} } undef, 0
  br label %18

18:                                               ; preds = %9, %15
  %"02.0" = phi { i32, {}, {} } [ %13, %9 ], [ { i32 0, {} undef, {} poison }, %15 ]
  ret { i32, {}, {} } %"02.0"
}


----- stderr -----
HUGR valid!
